### PR TITLE
OAS 3.0 - Use OAS 3.0 annotations instead of old Swagger 1.5.x / OAS 2.x ones (@Api / @ApiXxx)

### PR DESCRIPTION
**Pull request by reta**
As of now, Swagger Codegen always uses old (`1.5.x`) Swagger annotations, `@Api / @ApiModel` / `@ApiModelProperty / ...` to decorate model classes, APIs, operations, ... even if the input specification comes in OAS 3.0.x format. It would make sense to use the Swagger 2.0.x annotations in this case since it corresponds to OAS 3.0.x specification and has different, richer set of attributes. 

Issue: 
  - https://github.com/swagger-api/swagger-codegen-generators/issues/90

Server / client stubs updated:
 - [x] jaxrs-cxf-cdi
 - [x] jaxrs-spec
 - [x] jaxrs-cxf
 - [x] jaxrs-di
 - [x] jaxrs-jersey
 - [x] jaxrs-resteasy
 - [x] jaxrs-resteasy-eap
 - [x] client models (feign, ...)

As now `Springfox` (the Swagger/OpenAPI integration for `Spring` / `Spring Boot` ) does not support OpenAPI (Swagger 2.0.x), leaving it untouched.

Added new command-line argument to preserve the usages of the older OpenAPI 2.0 (Swagger 1.5.x) annotations and specification. 

From the command line:
```
java -jar modules/swagger-codegen-cli/target/swagger-codegen-cli.jar generate 
    -l java  
    -i http://petstore.swagger.io/v2/swagger.json  
    -o samples 
    --use-oas2 
```

From `Maven` plugin:
```
...
<configOptions>
    <useOas2>true</useOas2>
</configOptions>
...
```

### CODE REVIEW

**Comment by reta** [comment]
@webron @HugoMario Migrated all JAX-RS part, `Spring` seems to be behind due to `Springfox` (does not support Swagger 2.0.x yet). Please take a look guys, thanks!

File: src/main/java/io/swagger/codegen/languages/java/AbstractJavaCodegen.java [hunk 1]
```diff
@@ -145,6 +147,7 @@ public AbstractJavaCodegen() {
         cliOptions.add(CliOption.newBoolean(FULL_JAVA_UTIL, "whether to use fully qualified name for classes under java.util. This option only works for Java API client"));
         cliOptions.add(new CliOption(CodegenConstants.HIDE_GENERATION_TIMESTAMP, CodegenConstants.HIDE_GENERATION_TIMESTAMP_DESC));
         cliOptions.add(CliOption.newBoolean(WITH_XML, "whether to include support for application/xml content type and include XML annotations in the model (works with libraries that provide support for JSON and XML)"));
+        cliOptions.add(CliOption.newBoolean(CodegenConstants.USE_OAS2, "whether to use OpenAPI v2.0 (Swagger 1.5.x) or not (by default, OpenAPI v3.0 is used)"));
```
Comment (reta) [inline src/main/java/io/swagger/codegen/languages/java/AbstractJavaCodegen.java]: For use with `Maven` plugin

**Comment by reta** [comment]
@HugoMario Seems like we all set, if you mind taking a look please. Thanks!

**Comment by HugoMario** [comment]
thanks a lot @reta, going to look right now.

**Comment by HugoMario** [comment]
Hey @reta sorry for delay, i review code and looks pretty fine for me, but before merge i would like to check the generated output on `bin/*` files. i'll need a bit more time for it. Just letting you know that i'm working on this.

Also i'm not sure how contributor list works, but i saw that there is not email linked to your profile, this could cause that your PR doesn't count as contribution for Github once the PR is merged.

**Comment by reta** [comment]
Thanks a lot @HugoMario , really appreciate that. I have tested all codegens, no doubts, the compilation for server and client stubs was successful for both oas 2.0 / 3.0 but surely the bugs may escape. Regarding the contributor list, not sure how it works either, but not a big deal if I am not there, getting this feature working would be just great.  Please let me know if you discover some issues, I will help to fix them. Thanks!

**Comment by HugoMario** [comment]
thanks again @reta , pretty nice stuff !!!

**Comment by rockxwre** [comment]
@reta @HugoMario 
Maybe I am misunderstanding how it should work, but I am having difficulties using the `useOas2` option. In the following examples I use _jaxrs-jersey_ for the language argument.

First, I used the following command to list all possible options for the language:
```
java -jar swagger-codegen-cli.jar config-help -l jaxrs-jersey
```
which results in a list with options, including `useOas2`. This suggests that this option is supported.

When I use this command:
```
java -jar swagger-codegen-cli.jar generate -l jaxrs-jersey -i openapi.yaml --additional-properties useOas2=true
```
it looks like the option `useOas2` is ignored. The generated code is provided with Swagger 3 annotations, instead of Swagger 2.

The command:
```
java -jar swagger-codegen-cli.jar generate --use-oas2 -l jaxrs-jersey -i openapi.yaml
```
results in the error: _swagger-codegen: error: unrecognized arguments: '--use-oas2'_

When using the last command with _java_ as the language, it does work as expected: code is generated with Swagger 2 annotations.

Am I doing something wrong?

**Comment by reta** [comment]
Hey @rockxwre!

Your expectations about the feature are absolutely correct. I am curious what version of the Swagger Codegen you are using? Or you are building from source? The OAS2/OAS3 support had been merged into 3.x.x only.  

Thank you!

**Comment by rockxwre** [comment]
Hi @reta,

Thank you for your reply! I used version 3.0.11:

```
java -jar swagger-codegen-cli.jar version
3.0.11
```

**Comment by reta** [comment]
Sorry @rockxwre , took me awhile to trace the issue, indeed for command-line tooling the option is not accepted by all server generation libraries (but it is for all client libraries). I will try to submit a fix shortly but if you could use Swagger Codegen Maven Plugin (3.x) or Codegen 2.x (only OAS 2) meantime, it should just work fine. Thank you.

**Comment by SlavchoArsovski** [comment]
Any updates on this issue, the latest version still uses the old annotations instead of io.swagger.v3.oas.annotations

**Comment by reta** [comment]
@SlavchoArsovski haven't had a chance to work on it, will try to wrap it up with next 1-2 weeks, thanks for pinging.

**Comment by SlavchoArsovski** [comment]
@reta by any chance that swagger codegen for Spring and as Java Client will use io.swagger.v3.oas.annotation?
What you have added in the comment is for other languages:
x ] jaxrs-spec
[ x ] jaxrs-cxf
[ x ] jaxrs-resteasy-eap
[ x ] jaxrs-resteasy
[ x ] jaxrs-jersey
[ x ] jaxrs-cxf-cdi
[ x ] jaxrs-di


But not Spring Server stub and Java Client

**Comment by reta** [comment]
@SlavchoArsovski yes, the Spring Server stubs were merged in https://github.com/swagger-api/swagger-codegen-generators/pull/753, please give it a try, thanks! For Java Client, all of them should support `io.swagger.v3.oas`, which one you have in mind?

**Comment by SlavchoArsovski** [comment]
Hi @reta, i see on online editor when i generate API that indeed now that generated code contains  io.swagger.v3.oas.annotation, very cool!

However i have the maven plugin version: 3.0.22 and still generates the code with old swagger annotations, am i missing something?

**Comment by SlavchoArsovski** [comment]
Ahh i see i need to use 3.0.23, i will try it now 👍

**Comment by SlavchoArsovski** [comment]
@reta found 2 issues so for with migration from 3.0.21 to 3.032 ending with compile errors:

1. generateg SwaggerDocumentationConfig fails to compile becuase in SwaggerDocumentationConfig.mustache there is same class name from two different packages:
import io.swagger.v3.oas.models.info.Contact;
import springfox.documentation.service.Contact;

2. having default value on query parametar generates following code:
schema:

- name: 'offset'
        in: "query"
        description: 'Position of first item returned'
        required: false
        schema:
          type: integer
          format: int32
          example: 100
          default: 0
@Parameter(in = ParameterIn.QUERY, description = "Position of first item returned" ,schema=@Schema(, defaultValue="0")) @Valid @RequestParam(value = "offset", required = false, defaultValue="0") Integer offset 

---> schema=@Schema(, defaultValue="0")) has obsolete comma "," which makes the file with compile error

**Comment by reta** [comment]
@SlavchoArsovski could you please submit an issue to `https://github.com/swagger-api/swagger-codegen-generators`, I will try to pick it up right away, thank you.

**Comment by SlavchoArsovski** [comment]
@reta https://github.com/swagger-api/swagger-codegen-generators/issues/795

Do you know how can i exclude generation of SwaggerDocumentationConfig meanwhile, cause i do not really need it :)

**Comment by reta** [comment]
@S

> Do you know how can i exclude generation of SwaggerDocumentationConfig meanwhile, cause i do not really need it :)

Hm ... I think you could try these three things:
 - try to override the template  
 - or (could be easier), apply the code formatter to generated classes, the unused imports should be removed
 - remove the generated class for SwaggerDocumentationConfig altogether (after generation but before complilation)

**Pull request merged**